File: ts/problems/1-two-sum.test.ts
import { describe, expect, test } from "vitest";
import { twoSum } from "./1-two-sum";

describe("twoSum", () => {
  test.each([
    {
      nums: [],
      target: 0,
      want: [],
    },
    {
      nums: [2, 7, 11, 15],
      target: 9,
      want: [0, 1],
    },
    {
      nums: [3, 2, 4],
      target: 6,
      want: [1, 2],
    },
    {
      nums: [3, 3],
      target: 6,
      want: [0, 1],
    },
  ])(
    "from $nums with target of $target we want $want",
    ({ nums, target, want }) => {
      expect(twoSum(nums, target)).toStrictEqual(want);
    },
  );
});

File: go/problems/1_two_sum_test.go
package main

import (
	"fmt"
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestTwoSum(t *testing.T) {
	tests := []struct {
		nums   []int
		target int
		want   []int
	}{
		{
			nums:   []int{},
			target: 0,
			want:   []int{},
		},
		{
			nums:   []int{2, 7, 11, 15},
			target: 9,
			want:   []int{0, 1},
		},
		{
			nums:   []int{3, 3},
			target: 6,
			want:   []int{0, 1},
		},
	}

	for _, tt := range tests {
		t.Run(fmt.Sprintf("from %v with target of %d we want %v", tt.nums, tt.target, tt.want), func(t *testing.T) {
			got := twoSum(tt.nums, tt.target)
			assert.Equal(t, tt.want, got, "got %q want %q", got, tt.want)
		})
	}
}
